File: leetcode/01.two_sum/two_sum.py
######################
## Author: Duc Tran ##
## Title: Two Sum   ##
######################

class Solution:
    def twoSum(self, nums, target):
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i, j]

    def twoSum1(self, nums, target):
        values = {}
        for i, num in enumerate(nums):
            if (target - num) in values:
                return [values[target - num], i]
            values[num] = i
        return []

    # Sorted test
    def twoSum2(self, nums, target):
        i = 0
        j = len(nums) - 1

        while i < j:
            if nums[i] + nums[j] == target:
                return [i, j]
            elif nums[i] + nums[j] > target:
                j -= 1
            else:
                i +=1
        return []

if __name__ == '__main__':
    solution = Solution()
    nums = list([2, 7, 11, 15])
    target = 9
    print(solution.twoSum2(nums, target))

File: leetcode/01.two_sum/two_sum.go
//
// Author: Duc Tran
// Title: Two Sum
//

package main

import "fmt"

func twoSum(nums []int, target int) []int {
	results := []int{}
	for i := range nums {
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] == target {
				results = []int{i, j}
			}
		}
	}
	return results
}

func twoSum1(nums []int, target int) []int {
	m := map[int]int{}

	for i := range nums {
		temp := nums[i]
		search := target - temp

		if j, ok := m[search]; ok {
			return []int{i, j}
		}

		m[temp] = i
	}
	return []int{}
}

func main() {
	nums := []int{2, 7, 11, 15}
	target := 9
	fmt.Println(twoSum1(nums, target))
}
